addons/link_tracker/models/mail_render_mixin.py

```diff
@@ -35,24 +35,24 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
         """
         base_url = base_url or self.env['ir.config_parameter'].sudo().get_param('web.base.url')
         short_schema = base_url + '/r/'
-        for match in re.findall(tools.HTML_TAG_URL_REGEX, html):
-            href = markupsafe.Markup(match[0])
+        for match in set(re.findall(tools.HTML_TAG_URL_REGEX, html)):
             long_url = match[1]
-            label = (match[3] or '').strip()
+            if not (blacklist and any(s in long_url for s in blacklist) or long_url.startswith(short_schema)):
+                label = (match[3] or '').strip()
 
-            if not blacklist or not [s for s in blacklist if s in long_url] and not long_url.startswith(short_schema):
```

Comment:
Actually the motivation was initially the logic fault, as having a `False` `blacklist`  was enough to enter the block.
But I'll take the suggestion!

Reply:
Can't we do an early skip (aka: less negative conditions, less indent) and maybe split conditions ? Like
```
# already shortened url should not be shortened again
if long_url.startswith(short_schema):
    continue
# url present in blacklist (aka to skip list)
if blacklist and any(s in long_url for s in blacklist):
    continue
# do the processing
....
```
Seems it would make the code a bit easier to understand

EDIT: just saw it was done like that with text shortening, so basically I propose to make code a bit similar :)